include/envoy/http/header_map.h

```diff
@@ -65,9 +71,10 @@ class LowerCaseString {
   void lower() {
     std::transform(string_.begin(), string_.end(), string_.begin(), absl::ascii_tolower);
   }
-  bool valid() const { return validHeaderString(string_); }
+  bool valid() const { return validHeaderString(string_) && validHeaderString(original_string_); }
 
   std::string string_;
+  std::string original_string_;
```

Comment:
They're not gonna accept duplicate strings for _every_ header key regardless of whether case preservation is opt in. They mention in this file that they optimize for performance. You need to either not populate this member unless case preservation is opt-in, or you need to remove this member altogether and stop populating on LowerCaseString,